src/ts/services/TelemetryService.ts

```diff
@@ -1,69 +1,55 @@
-    this.sendTelemetryEvent({
-      type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_SUGGESTION_IS_ACCEPTED,
-      value: 1,
-      eventTime: new Date().toISOString(),
-      tags
-    });
-  }
+class UserTelemetryEventSender {
+    private postEventLimit = 500;
+    private eventBuffer: ITyperighterTelemetryEvent[] = [];
 
-  public matchIsMarkedAsCorrect(tags: IMarkAsCorrectEvent["tags"]) {
-    this.sendTelemetryEvent({
-      type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_MARK_AS_CORRECT,
-      value: 1,
-      eventTime: new Date().toISOString(),
-      tags
-    });
-  }
+    public constructor(private telemetryUrl: string, private throttleDelay: number = 10000) {};
 
-  public documentIsChecked(tags: ITyperighterTelemetryEvent["tags"]) {
-    this.sendTelemetryEvent({
-      type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_CHECK_DOCUMENT,
-      value: 1,
-      eventTime: new Date().toISOString(),
-      tags
-    });
-  }
+    private async sendEvents(): Promise<void> {
+        const [firstChunk, ...subsequentChunks] = chunk(this.eventBuffer, this.postEventLimit);
+        if (!firstChunk) {
+          return Promise.resolve();
+        }
 
-  public typerighterIsOpened(tags: ITyperighterTelemetryEvent["tags"]) {
-    this.sendTelemetryEvent({
-        type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_OPEN_STATE_CHANGED,
-        value: true,
-        eventTime: new Date().toISOString(),
-        tags
-    });
-  }
+        const jsonEventBuffer = JSON.stringify(firstChunk);
 
-  public typerighterIsClosed(tags: ITyperighterTelemetryEvent["tags"]) {
-    this.sendTelemetryEvent({
-        type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_OPEN_STATE_CHANGED,
-        value: false,
-        eventTime: new Date().toISOString(),
-        tags
-    });
-  }
+        // Push the remaining events back into the buffer
+        this.eventBuffer = subsequentChunks.flat();
 
-  public sidebarMatchClicked(tags: ISidebarClickEvent["tags"]) {
-    this.sendTelemetryEvent({
-        type: TYPERIGHTER_TELEMETRY_TYPE.TYPERIGHTER_SIDEBAR_MATCH_CLICK,
-        value: 1,
-        eventTime: new Date().toISOString(),
-        tags
-    });
-  }
+        const response = await fetch(`${this.telemetryUrl}/event`, {
+          method: "POST",
+          mode: "cors",
+          credentials: "include",
+          headers: new Headers({
+            "Content-Type": "application/json"
+          }),
+          body: jsonEventBuffer
+        });
+
+        if (!response.ok) {
+          this.eventBuffer = this.eventBuffer.concat(firstChunk);
+        }
+
+        if (this.eventBuffer.length) {
+
+          this.throttledSendEvents();
+        }
+    }
+
+    private throttledSendEvents = throttle(this.sendEvents, this.throttleDelay, { trailing: true, leading: false })
+
+    public addEvent(event: ITyperighterTelemetryEvent): void {
+      this.eventBuffer.push(event);
+      this.throttledSendEvents();
+    }
+
+    public flushEvents(): Promise<void> {
```

Comment:
Is this going to be wired up in a subsequent PR?

Reply:
It's not really necessary in our page demo – we've tests for this, so I'm not too worried, but we could wire this up to a button or something to test if we need to.